cpp/src/unary/math_ops.cu

```diff
@@ -0,0 +1,428 @@
+    cudf::experimental::unary::launcher<T, T, F>::launch(input, output);
+}
+
+
+template <typename F>
+struct MathOpDispatcher {
+    template <typename T>
+    typename std::enable_if_t<std::is_arithmetic<T>::value, void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        launch<T, F>(input, output);
+    }
+
+    template <typename T>
+    typename std::enable_if_t<!std::is_arithmetic<T>::value, void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        CUDF_FAIL("Unsupported datatype for operation");
+    }
+};
+
+
+template <typename F>
+struct BitwiseOpDispatcher {
+    template <typename T>
+    typename std::enable_if_t<std::is_integral<T>::value, void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        launch<T, F>(input, output);
+    }
+
+    template <typename T>
+    typename std::enable_if_t<!std::is_integral<T>::value, void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        CUDF_FAIL("Unsupported datatype for operation");
+    }
+};
+
+
+template <typename F>
+struct LogicalOpDispatcher {
+private:
+    template <typename T>
+    static constexpr bool is_supported() {
+        return std::is_arithmetic<T>::value ||
+               std::is_same<T, cudf::experimental::bool8>::value;
+
+        // TODO: try using member detector
+        // std::is_member_function_pointer<decltype(&T::operator!)>::value;
+    }
+
+public:
+    template <typename T>
+    typename std::enable_if_t<is_supported<T>(), void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        cudf::experimental::unary::launcher<T, cudf::experimental::bool8, F>::launch(input, output);
+    }
+
+    template <typename T>
+    typename std::enable_if_t<!is_supported<T>(), void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        CUDF_FAIL("Unsupported datatype for operation");
+    }
+};
+
+} // namespace detail
+
+std::unique_ptr<cudf::column>
+unary_operation(cudf::column_view const& input,
+                cudf::experimental::unary_op op,
+                cudaStream_t stream,
+                rmm::mr::device_memory_resource* mr) {
+
+    std::unique_ptr<cudf::column> output = [&] {
+        if (op == cudf::experimental::unary_op::NOT) {
+
+            auto type = cudf::data_type{cudf::BOOL8};
+            auto size = input.size();
+
+            return std::make_unique<column>(
+                type, size,
+                rmm::device_buffer{size * cudf::size_of(type), 0, mr},
+                copy_bitmask(input, 0, mr),
+                input.null_count());
+
+        } else {
+            return cudf::experimental::allocate_like(input);
+        }
+    } ();
+
+    if (input.size() == 0) return output;
+
+    auto output_view = output->mutable_view();
```

Comment:
This pattern of passing in the output as a mutable_view prevents any future support of string or other non fixed-width type columns. Granted, most of these unary ops don't make sense on strings. However, I believe these should be modified to _return_ their output rather than pass via a mutable view.

Reply:
dispatcher / launcher is updated to return `std::unique_ptr<cudf::column_view>`